peer_heap.js

```diff
@@ -40,69 +46,62 @@ PeerHeap.prototype.choose = function choose(threshold, filter) {
         return null;
     }
 
-    var el;
-    if (filter) {
-        el = self._chooseFilteredEl(threshold, filter);
-    } else {
-        el = self._chooseEl(threshold);
-    }
-
-    if (!el) {
-        return null;
-    }
-
-    return el.peer;
-
-};
-
-PeerHeap.prototype._chooseEl = function _chooseEl(threshold) {
-    var self = this;
-
-    var el = self.array[0];
-    if (el.score <= threshold) { // TODO: why inclusive?
-        return null;
-    }
-
-    return el;
-};
-
-PeerHeap.prototype._chooseFilteredEl = function _chooseFilteredEl(threshold, filter) {
-    var self = this;
-
-    // TODO: is it worth it to unroll the first iteration of the loop below so
-    // that we incur minimal overhead for "the top of heap is okay" case?
-
-    // NOTE: assumes self._stack starts off empty
-    self._stack.push(0);
-    while (self._stack.length) {
-        var i = self._stack.shift();
+    var chosenPeer = null;
+    var maxRangeStart = self.array[0].range.lo;
+    var highestProbability = 0;
+    var firstScore = self.array[0].peer.getScore();
+    var stackBegin = 0;
+    var stackEnd = 0;
+
+    if (firstScore > threshold && (!filter || filter(self.array[0].peer))) {
+        // Don't check first peer if it looks good, check its children though
+        chosenPeer = self.array[0].peer;
+        highestProbability = firstScore;
+        stackBegin = 1;
+        stackEnd = 2;
+    } 
+
+    while (stackBegin <= stackEnd) {
+        var i = dfsStack[stackBegin];
+        stackBegin++;
 
         var el = self.array[i];
-        if (el.score <= threshold) { // TODO: why inclusive?
-            break;
-        }
 
-        if (!filter || filter(el.peer)) {
-            return el;
+        if (self.rangehis[i] <= maxRangeStart) {
+            // This range ends before the range with the largest start begins,
+            // so it can't possibly be chosen over any of the ranges we've
+            // seen. All ranges below this one have a smaller end, so this
+            // range and any below it can't be chosen.
+            continue;
+        } else if (!filter || filter(el.peer)) {
+            maxRangeStart = Math.max(maxRangeStart, self.rangelos[i]);
+
+            // INLINE of TChannelPeer#getScore
+            var lo = self.rangelos[i];
```

Comment:
Yes, we can do that. It has a huge improvement because it removes the need for the range hi/lo property lookups. Instead, they're in a contiguous array, which improves cache locality.